### PR TITLE
[SCB-149] service aware callback

### PR DESCRIPTION
**Pull request by seanyinx**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  public CompositeOmegaCallback(Map<String, Map<String, OmegaCallback>> callbacks) {
+    this.callbacks = callbacks;
+  }
+
+  @Override
+  public void compensate(TxEvent event) {
+    Map<String, OmegaCallback> serviceCallbacks = callbacks.get(event.serviceName());
+
+    if (serviceCallbacks.isEmpty()) {
+      throw new AlphaException("No such omega callback found for service " + event.serviceName());
```
Comment (WillemJiang) [inline alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: Here we just know about the service name, I think we need to put the event here.

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    OmegaCallback omegaCallback = serviceCallbacks.get(event.instanceId());
+    if (omegaCallback == null) {
+      serviceCallbacks.values().iterator().next().compensate(event);
```
Comment (WillemJiang) [inline alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: It looks like a default compensate call back here.  How about there are different instance of service which need to be compensated?

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/RetryOmegaCallback.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  private final OmegaCallback underlying;
+  private final int delay;
+
+  public RetryOmegaCallback(OmegaCallback underlying, int delay) {
+    this.underlying = underlying;
+    this.delay = delay;
+  }
+
+  @Override
+  public void compensate(TxEvent event) {
```
Comment (WillemJiang) [inline alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/RetryOmegaCallback.java]: The calling thread is blocked, we may need to think for another way to do it.

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java (hunk 1 above)
Comment (seanyinx) [reply alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: what event?

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java (hunk 2 above)
Comment (seanyinx) [reply alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: there are multiple instances of the same service stored in the map, we just choose the first instance, if the matching instance is unreachable and removed from the map

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/RetryOmegaCallback.java (hunk 3 above)
Comment (seanyinx) [reply alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/RetryOmegaCallback.java]: the compensation is done in a separate thread from the event handling thread.

File: alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java (hunk 1 above)
Comment (WillemJiang) [reply alpha/alpha-core/src/main/java/io/servicecomb/saga/alpha/core/CompositeOmegaCallback.java]: How about add the event to the exception, or log the event.

**Pull request merged**